### ISSUE TITLE
[feature] Expose `self.dependencies` in all methods of conanfile called after `validate()`

### ISSUE THREAD

**Issue by SpaceIm** [conan-io/conan#11940]
- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

In https://docs.conan.io/en/1.51/reference/conanfile/dependencies.html, there is:

> At the moment, this information should only be used in `generate()` and `validate()` methods. Any other use, please submit a Github issue.

So I'm submitting an issue to expose self.dependencies :)

Actually in this PR of `tinyspline` recipe https://github.com/conan-io/conan-center-index/pull/12323, I need to check an option value of tinyspline in test() method of test_package, and it fails with:

```
ERROR: tinyspline/0.3.0 (test package): Error in test() method, line 31
	if self.dependencies["tinyspline"].options.cxx:
	AttributeError: 'NoneType' object has no attribute 'dependencies'
```

https://c3i.jfrog.io/c3i/misc/logs/pr/12323/4-linux-gcc/tinyspline/0.3.0//ab460bcf84a19ada9f4e0bfddca5324c1490a3ea-test.txt

self.dependencies should be exposed at least in `validate()`, `validate_build()`, `build()`, `package()`, `package_info()`, `test()`.

**Comment by memsharded** [comment]
> Actually in this PR of tinyspline recipe https://github.com/conan-io/conan-center-index/pull/12323, I need to check an option value of tinyspline in test() method of test_package, and it fails with:

This is impossible in 1.X, as the ``test_package->test()`` method is local, and it does not compute the dependency graph, so it cannot have ``self.dependencies``. Whatever you need to do it, use the ``generate()`` method of ``test_package/conanfile.py``, to save some files, and recover that information in the ``test()`` method.

**Comment by SpaceIm** [comment]
Ok, I'll find a workaround.

Does it mean than, except for `test()`, `self.dependencies` already works properly in other methods than `validate()` & `build()`, and documentation needs an update to list supported methods (and eventually if it needs CONAN_V2 mode or not)?

**Comment by memsharded** [comment]
``self.dependencies`` will be in post-graph built methods:
- ``generate()``
- ``validate()``
- ``package_id()``
- ``package_info()``
It will not be in pre-graph methods (not now, not in Conan 2.0, it is impossible by construction):
- ``configure()``
- ``config_options()``
- ``requirements()``, ``build_requirements()``
- ``init()``
It is not in methods that can execute locally (``build()``, ``test()``, and ``package()`` will have it in 2.0, but still not documented):
- ``source()``
- ``build()``
- ``test()``
- ``package()``

So the current documentation seems correct. ``self.dependencies`` should only be used in ``generate()`` and ``validate()``. If necessary for some reason in ``package_id()`` or ``package_info()`` those are the cases we could consider to document. For the ``build()``, ``test()``, and ``package()``, we will reconsider after 2.0)

**Comment by SpaceIm** [comment]
Thanks, it's a useful information for conan v2 migration.

Isn't `build_requirements()` a post host graph built method?

**Comment by memsharded** [comment]
> Isn't build_requirements() a post host graph built method?

No, it is resolved while resolving the graph (the graph also includes build-requires). ``self.dependencies`` cannot and shouldn't be used inside ``build_requirements()``

**Comment by franramirez688** [comment]
Should this issue be linked to Conan 2.x? Or should we open another one? Or even close this one?
